Drivers: ssd1306: Support for EastRising 0.42 OLED

theob-pro opened the pull request:
Add `use_internal_iref` DTS option, this is necessary for supporting EastRising 0.42 OLED display/board.

Some boards don't have external Iref set up. This is probably done in an effort to save on component cost. This command is only documented in the V1.1 revision of the SSD1306 datasheet.

See issue https://github.com/olikraus/u8g2/issues/1047

Also update `ssd1306_write_default` to take into account the `segment_offset` property. This is needed for some displays to show the image correctly.

theob-pro committed:
Drivers: ssd1306: Add use_internal_iref DTS option

Necessary for supporting for EastRising 0.42 OLED display/board.

Some boards don't have external Iref set up. This is probably done in an
effort to save on component cost. This command is only documented in the
V1.1 revision of the SSD1306 datasheet.

See issue https://github.com/olikraus/u8g2/issues/1047

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

decsny commented on dts/bindings/display/solomon,ssd1306fb-common.yaml:
```diff
@@ -60,3 +60,13 @@ properties:
+    description: |
+      Use internal Iref, not common but sometime used to save components cost
+
+  use-alternative-write:
```
why cant this be kconfig

narvalotech replied:
what's the rationale, why can't the other dts properties be kconfig?

danieldegrasse replied:
I'm confused about your comment here:

> The write command for ssd1306 doesn't take into account the `segment_offset` DTS property. However the one used for ht1306 does. Allow
users to use that write method on non-ht1306 controllers.

Could we instead update the ssd1306 write logic to account for the `segment_offset` property?

jori-nordic replied:
There are two write functions in the ssd1306 driver. 
Currently, which to use is decided based on the `compatible =` dts property. If it's `ssd1306`, the first one is used, it's `ht1306`, then the second one is used.

The particular display that the board in the linked PR has is an ssd1306, but has a segment offset that needs to be set to display the image correctly.

The second write method (originally intended for the ht1306, but that's not what we have) is the only one that works with this display.

We didn't want to touch the other write method as we don't have other hardware than this, and can't test that we don't introduce a regression.

jori-nordic replied:
I didn't realize I did this, but I also commented with @narvalotech which is my personal account. Sorry bout that :)

decsny replied:
Please put more info in the description of the property in the binding so thats users of the binding know when or not to use it.

theob-pro replied:
We updated the `ssd1306_write_default` instead.

danieldegrasse replied:
I'm concerned this would introduce a regression as well, since it is only tested on this display. Could this display use the `ht1306` compatible? I'm not clear what the differences between these displays are. Otherwise perhaps we can test this with another display to check for regressions, @henrikbrixandersen do you have access to a display using this driver? I see you added it to a board in #70645

jori-nordic replied:
It uses at least one command that is not listed in the `ht1306` command set / datasheet. I think it's the Iref command.

danieldegrasse replied:
Ok- from the datasheet it does seem like the offset should be used there- and no in tree display definitions for the SSD1306 that I can find are setting the segment offset to a nonzero value- so I think this change is ok.

theob-pro committed:
Drivers: ssd1306: Respect `segment-offset` DTS property

Update `ssd1306_write_default` to take into account the `segment_offset`
property. This is needed for some displays to show the image correctly.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

danieldegrasse reviewed (commented):
@theob-pro could you add more details on the Iref property to address the change request? From what I can find online, the command is not in the datasheet for the SSD1306, but it does appear in the datasheet for the SSD1316: https://www.orientdisplay.com/wp-content/uploads/2022/08/SSD1316.pdf. Is it possible this display uses a SSD1316 controller, and we should use a different compatible?

theob-pro commented:
> @theob-pro could you add more details on the Iref property to address the change request? From what I can find online, the command is not in the datasheet for the SSD1306, but it does appear in the datasheet for the SSD1316: https://www.orientdisplay.com/wp-content/uploads/2022/08/SSD1316.pdf. Is it possible this display uses a SSD1316 controller, and we should use a different compatible?

From https://github.com/olikraus/u8g2/issues/1047#issuecomment-554624908 :

> None of the libraries I've looked at cater for this - maybe because this
feature is only documented in SSD1306B v1.1 data sheet and not in SSD1306
v1.

[![Screenshot from 2024-07-22 13-22-57](https://github.com/user-attachments/assets/2ce11e15-0efe-41fb-967c-2dd94f85d566)](http://buydisplay.com/download/ic/SSD1306.pdf)

I am not sure what you want me to add, could you suggest something?

jori-nordic commented:
>Please put more info in the description of the property in the binding

@decsny The property doesn't exist anymore. Could you remove your change-request?

decsny reviewed (approved):
binding change fine
